src/util/hb_arena.c

```diff
@@ -3,60 +3,152 @@
 #endif
 
 #include "../include/util/hb_arena.h"
+#include "../include/macros.h"
 
+#include <assert.h>
 #include <stdbool.h>
+#include <string.h>
 #include <sys/mman.h>
 
-#define KB(kb) (1024 * kb)
-#define MB(mb) (1024 * KB(mb))
+#define hb_arena_for_each_page(allocator, page)                                                                        \
+  for (hb_arena_page_T* page = (allocator)->head; page != NULL; page = page->next)
 
-bool hb_arena_init(hb_arena_T* allocator, size_t size) {
-  allocator->memory = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+static inline size_t hb_arena_align(size_t size, size_t alignment) {
+  return ((size + (alignment - 1)) / alignment) * alignment;
+}
+
+static inline bool hb_arena_page_is_full(hb_arena_page_T* page, size_t required_size) {
+  return page->position + required_size > page->capacity;
+}
 
-  if (allocator->memory == MAP_FAILED) {
-    allocator->memory = NULL;
-    allocator->position = 0;
-    allocator->capacity = 0;
+static inline void* hb_arena_page_alloc_from(hb_arena_page_T* page, size_t size) {
+  void* result = &page->memory[page->position];
+  page->position += size;
 
-    return false;
+  return result;
+}
+
+static inline void hb_arena_page_reset(hb_arena_page_T* page) {
+  page->position = 0;
+}
+
+static inline void hb_arena_reset_pages_after(hb_arena_page_T* start_page) {
+  for (hb_arena_page_T* page = start_page; page != NULL; page = page->next) {
+    hb_arena_page_reset(page);
   }
+}
+
+static bool hb_arena_append_page(hb_arena_T* allocator, size_t minimum_size) {
+  size_t page_size = allocator->default_page_size;
+  if (minimum_size > page_size) { page_size = minimum_size; }
+
+  size_t total_size = sizeof(hb_arena_page_T) + page_size;
+  hb_arena_page_T* page = mmap(NULL, total_size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+
+  if (page == MAP_FAILED) { return false; }
 
-  allocator->position = 0;
-  allocator->capacity = size;
+  *page = (hb_arena_page_T) { .next = NULL, .capacity = page_size, .position = 0 };
+
+  if (allocator->head == NULL) {
+    allocator->head = page;
+  } else {
+    allocator->tail->next = page;
+  }
+
+  allocator->tail = page;
 
   return true;
 }
 
+bool hb_arena_init(hb_arena_T* allocator, size_t initial_size) {
+  if (initial_size == 0) { return false; }
+
+  allocator->head = NULL;
+  allocator->tail = NULL;
+  allocator->default_page_size = initial_size;
+
+  return hb_arena_append_page(allocator, initial_size);
+}
+
 void* hb_arena_alloc(hb_arena_T* allocator, size_t size) {
-  size_t required_size = ((size + 7) / 8) * 8; // rounds up to the next 8 bytes
+  assert(size > 0);
 
-  if (allocator->position + required_size > allocator->capacity) { return NULL; }
+  size_t required_size = hb_arena_align(size, 8);
```

Comment:
Suggestion: Rename to fit actual behavior 

Very important, `hb_arena_align` does not align the returned pointer but the **next pointer**.

You add more space than you need at the end, so that the next pointer is aligned to i.e. 8 bytes.

```
+-----------------------------------+----------------------------------- 
| alloc (4bytes)  | 4 bytes padding | NEXT ....
+-----------------------------------+-----------------------------------
^
|
This is the memory pointer we are going to return
```

See [`hb_arena_page_alloc_from`](https://github.com/marcoroth/herb/pull/708/files#diff-38530fcdce39968fc659335346b4b6c0d6be9af0f16e2ed1354ca3bb693b2461R24)